File: src/javascript/algorithms/dynamicProgramming/climbingStairs.js
// TC: O(n) SC: O(1)
function climbStairs1(n) {
    if(n ===1 || n === 2) return 1;

    let first = 1, second = 1;
    let temp = 0;
    for(let i=0; i<n-1; i++) {
        temp = first + second;
        first = second;
        second = temp;
    }
    return second;
}

// TC: O(n) SC: O(n)
function climbStairs2(n) {
    if(n ===1 || n === 2) return 1;

    let dp = new Array(n+1);
    dp[0] = 1, dp[1] = 1;
    for(let i=2; i<n+1; i++) {
        dp[i] = dp[i-1] + dp[i-2];
    }
    return dp[n];
}

console.log(climbStairs1(6));
console.log(climbStairs2(6));

File: src/java1/algorithms/dynamicProgramming/climbingStairs/ClimbingStairs.java
package java1.algorithms.dynamicProgramming.climbingStairs;

public class ClimbingStairs {
    //DP Two variables TC: O(n) SC: O(1)
    private static int climbStairs1(int n) {
        if(n <= 2) return n;

        int first = 1, second = 1, temp;
        for(int i=2; i<= n; i++) { //n-1 times
            temp = first + second;
            first = second;
            second = temp;
        }
        return second;
    }

    //Bottom-up DP  TC:O(n) SC: O(n)
    private static int climbStairs2(int n) {
        if(n <= 2) return n;

        int[] dp = new int[n+1];
        dp[0] = 1; 
        dp[1] = 1;
        for(int i=2; i<=n; i++) {
            dp[i] = dp[i-1] + dp[i-2];
        } 
        return dp[n];
    }

    //Recursive approach TC: O(2^n) SC: O(n)
    private static int climbStairs3(int n) {
        if(n <=2) return n;

        return climbStairs3(n-1) + climbStairs3(n-2);
    }

    public static void main(String[] args) {
        System.out.println(climbStairs1(6));
        System.out.println(climbStairs2(6));
        System.out.println(climbStairs3(6));
    }
}
